addon/components/polaris-text-field.js

```diff
@@ -0,0 +1,520 @@
+   */
+  maxLength: null,
+
+  /**
+   * Mimics the behavior of the native HTML attribute,
+   * limiting how low the spinner can decrement the value
+   *
+   * @property min
+   * @public
+   * @type {Number}
+   * @default null
+   */
+  min: null,
+
+  /**
+   * Minimum character length for an input
+   *
+   * @property minLength
+   * @public
+   * @type {Number}
+   * @default null
+   */
+  minLength: null,
+
+  /**
+   * A regular expression to check the value against
+   *
+   * @property pattern
+   * @public
+   * @type {String}
+   * @default null
+   */
+  pattern: null,
+
+  /**
+   * Indicate whether value should have spelling checked
+   *
+   * @property spellCheck
+   * @public
+   * @type {Boolean}
+   * @default null
+   */
+  spellCheck: null,
+
+  /**
+   * Adds an action to the label
+   *
+   * @property labelAction
+   * @public
+   * @type {Object}
+   * @default null
+   *
+   * Currently supports:
+   * { onAction, content, accessibilityLabel }
+   */
+  labelAction: null,
+
+  /**
+   * Callback when value is changed
+   *
+   * @property onChange
+   * @public
+   * @type {Function}
+   * @default noop
+   */
+  onChange(/* value, id */) {},
+
+  /**
+   * Callback when input is focused
+   *
+   * @property onFocus
+   * @public
+   * @type {Function}
+   * @default noop
+   */
+  onFocus() {},
+
+  /**
+   * Callback when focus is removed
+   *
+   * @property onBlur
+   * @public
+   * @type {Function}
+   * @default noop
+   */
+  onBlur() {},
+
+  addValueListener() {
+    let [ field ] = this.element.querySelectorAll('input, textarea');
+    let _this = this;
+
+    field.addEventListener('keyup', () => {
+      debounceTask(_this, 'debouncedUpdateValue', 250);
+    });
+  },
+
+  removeValueListener() {
+    let [ field ] = this.element.querySelectorAll('input, textarea');
+
+    // inline method is passed so tests don't fail
+    let _this = this;
+    field.removeEventListener('keyup', () => {
+      debounceTask(_this, 'debouncedUpdateValue', 250);
+    });
+  },
+
+  debouncedUpdateValue() {
+    let [ field ] = this.element.querySelectorAll('input, textarea');
+    this.set('value', field.value);
+  },
+
+  textFieldClasses: computed('value', 'disabled', 'readOnly', 'error', 'multiline', 'focus', function() {
+    let { value, disabled, readOnly, error, multiline, focus } = this.getProperties('value', 'disabled', 'readOnly', 'error', 'multiline', 'focus');
+
+    let valueClass = value ? 'Polaris-TextField--hasValue' : '';
+    let disabledClass = disabled ? 'Polaris-TextField--disabled' : '';
+    let readOnlyClass = readOnly ? 'Polaris-TextField--readOnly' : '';
+    let errorClass = error ? 'Polaris-TextField--error' : '';
+    let multilineClass = multiline ? 'Polaris-TextField--multiline' : '';
+    let focusClass = focus ? 'Polaris-TextField--focus' : '';
+
+    return `Polaris-TextField ${ valueClass } ${ disabledClass } ${ readOnlyClass } ${ errorClass } ${ multilineClass } ${ focusClass }`;
```

Comment:
lol, this gives you one long chunk of trailing whitespace if none of those flags are truthy 😛 How do you feel about pushing the classes onto an array and returning `classes.join(' ')`?

Reply:
You'd still get the `Polaris-TextField` class though 😄 

I'm good with `.join(' ' )`, I'll update it.